index.js

```diff
@@ -41,14 +45,39 @@ exports.isPoweredOn = () => {
 };
 
 /** Brightness managment */
-exports.getBrightness = () => {
-    return readValue('actual_brightness').then((brightnessValue) => parseInt(brightnessValue, 10));
+exports.getBrightness = (percentage = true) => {
+    return  getMaxBrightness().then((maxBrightnessValue) => readValue('actual_brightness').then((brightnessValue) => {
+        return percentage ? Math.round((parseInt(brightnessValue, 10) * 100) / maxBrightnessValue) : parseInt(brightnessValue, 10);
+    }));
 };
 
-exports.setBrightness = (value) => {
-    return this.getMaxBrightness().then((maxBrightnessValue) => writeValue('brightness', value));
+exports.setBrightness = (value, speed = 0) => {
+    return getMaxBrightness().then((maxBrightnessValue) =>  {
+        let toValue = Math.round((parseInt(value, 10) * maxBrightnessValue) / 100);
+        speed = parseInt(speed, 10);    
+        if (speed > 0 && toValue > 0) {
+            return this.getBrightness(false).then((actualBrightnessValue) => {
+                if (toValue < actualBrightnessValue) {
+                    let goDown = setInterval(() => { 
+                        if (actualBrightnessValue > toValue) {
+                            writeValue('brightness', actualBrightnessValue--); 
+                        } else {
+                            clearInterval(goDown);
+                        }                   
+                    }, speed);
+                } else if (toValue > actualBrightnessValue) {
+                    let goUp = setInterval(() => {
+                        if (actualBrightnessValue < toValue) {
+                            writeValue('brightness', actualBrightnessValue++); 
+                        } else {
+                            clearInterval(goUp);
+                        }                   
+                    }, speed);
```

Comment:
This animating logic is pretty complicated, doesn't quite handle promises right (the promise resolves as soon as you've read the current brightness, but should really wait until the animation is finished), has some odd behaviour in a few edge cases (what happens if you run two animations at once, or if you call set while one is running?), and for some reason doesn't animate if you're going to 0?

At the least I think we need to split this into a separate function from `setBrightness` generally (e.g. `animateBrightness`), and refactor this logic inside too: you should be able to separate the 'run a function with an increasing value over time' logic out entirely from the 'change the brightness' logic. That would remove the duplication here, generally make this easier to follow, and make some of the above easier to solve.

Hmm, that does actually suggest that we should just delegate the timing part to one of the many separate libraries that already do this though. We can do that pretty easily actually, if we install [`TinyAnimate` ](https://www.npmjs.com/package/TinyAnimate), this should be something like:

```javascript
return new Promise((resolve) => {
  currentAnimation = TinyAnimate.animate(currentBrightness, targetBrightness, duration, setBrightness, 'linear', resolve)
});
```

(I think, I haven't actually tried this).

From there, with just a couple of extra lines you can use `currentAnimation` to cancel the running animation if you start a new one or change the value some other way, etc etc. That's just the first library I found with a few minutes search, but I'm open to basically any small lib that does this, bonus points if it already does promises natively (but most of them don't seem to, sadly). You could also easily add easing arguments to this new `animateBrightness` function, passing them directly into `TinyAnimate`, so you can do it smoothly or bounce or whatever you like: http://easings.net/.

How does that sound?